operator/pkg/lbipam/range_store.go

```diff
@@ -0,0 +1,150 @@
+	if idx != -1 {
+		rs.ranges = slices.Delete(rs.ranges, idx, idx+1)
+	}
+
+	poolRanges := rs.poolToRanges[lbRange.originPool]
+
+	idx = slices.Index(poolRanges, lbRange)
+	if idx != -1 {
+		poolRanges = slices.Delete(poolRanges, idx, idx+1)
+	}
+
+	if len(poolRanges) > 0 {
+		rs.poolToRanges[lbRange.originPool] = poolRanges
+	} else {
+		delete(rs.poolToRanges, lbRange.originPool)
+	}
+}
+
+func (rs *rangesStore) Add(lbRange *LBRange) {
+	rs.ranges = append(rs.ranges, lbRange)
+	poolRanges := rs.poolToRanges[lbRange.originPool]
+	poolRanges = append(poolRanges, lbRange)
+	rs.poolToRanges[lbRange.originPool] = poolRanges
+}
+
+func (rs *rangesStore) GetRangesForPool(name string) ([]*LBRange, bool) {
+	ranges, found := rs.poolToRanges[name]
+	return ranges, found
+}
+
+type LBRange struct {
+	// the actual data of which ips have been allocated or not
+	allocRange *ipallocator.Range
+	// If true, the LB range has been disabled via the CRD and thus no IPs should be allocated from this range
+	externallyDisabled bool
+	// If true, the LB range has been disabled by us, because it conflicts with other ranges for example.
+	// This range should not be used for allocation.
+	internallyDisabled bool
+	// The name of the pool that originated this LB range
+	originPool string
+}
+
+func NewLBRange(cidr *net.IPNet, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) (*LBRange, error) {
+	allocRange, err := ipallocator.NewCIDRRange(cidr)
+	if err != nil {
+		return nil, fmt.Errorf("new cidr range: %w", err)
+	}
+
+	return &LBRange{
+		allocRange:         allocRange,
+		internallyDisabled: false,
+		externallyDisabled: pool.Spec.Disabled,
+		originPool:         pool.GetName(),
+	}, nil
+}
+
+func (lr *LBRange) Disabled() bool {
+	return lr.internallyDisabled || lr.externallyDisabled
+}
+
+func (lr *LBRange) String() string {
+	cidr := lr.allocRange.CIDR()
+	return fmt.Sprintf(
+		"%s (free: %d, used: %d, intDis: %v, extDis: %v) - origin %s",
+		cidr.String(),
+		lr.allocRange.Free(),
+		lr.allocRange.Used(),
+		lr.internallyDisabled,
+		lr.externallyDisabled,
+		lr.originPool,
+	)
+}
+
+func ipNetStr(net net.IPNet) string {
+	ptr := &net
+	return ptr.String()
+}
+
+// areRangesInternallyConflicting checks if any of the ranges within the same list conflict with each other.
+func areRangesInternallyConflicting(ranges []*LBRange) (conflicting bool, rangeA, rangeB *LBRange) {
+	for i, outer := range ranges {
+		for ii, inner := range ranges {
+			if i == ii {
+				continue
+			}
+
+			if !intersect(outer.allocRange.CIDR(), inner.allocRange.CIDR()) {
+				continue
+			}
+
+			return true, outer, inner
+		}
+	}
+
+	return false, nil, nil
+}
+
+func areRangesConflicting(outerRanges, innerRanges []*LBRange) (conflicting bool, targetRange, conflictingRange *LBRange) {
+	for _, outerRange := range outerRanges {
+		for _, innerRange := range innerRanges {
+			// IPs of dissimilar IP families can't overlap
+			outerIsIpv4 := outerRange.allocRange.CIDR().IP.To4() != nil
```

Comment:
Maybe add a convenience Is4() function to `range`? Then the skip is just `outerRange.Is4() != innerRange.Is4()`. Just a cosmetic thing.